{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bio 302 Homework 8\n",
    "\n",
    "**Due before class Thursday, April 21st**\n",
    "\n",
    "This homework consists of the two problems in this notebook. You may find the reading in Chapter XX useful background.\n",
    "\n",
    "**Submission**\n",
    "\n",
    "Submit your solutions as a link to a Gist via a Slack DM."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem 1\n",
    "\n",
    "Recall that diploid organisms can have two kinds of genotypes on their\n",
    "autosomal chromosomes: *homozygotes* and *heterozygotes*. You're\n",
    "homozygous at a locus if you got the same allele from both mom and\n",
    "dad, *e.g.* `A/A`; you're heterozygous at a locus if you have\n",
    "different alleles, *e.g.* `A/G`.\n",
    "\n",
    "Runs of homozygosity (ROHs) are longer-than-expected runs of\n",
    "consecutive homozygotes along a chromosome. They can occur for a\n",
    "number of reasons, and are most common in culturally or geographically\n",
    "isolated populations. ROHs may lead to poor health, since they have\n",
    "the potential to expose recessive disease. For more on the connections\n",
    "between ROHs, history and health, you can check out the following:\n",
    "@nalls_measures_2009, @kirin_genomic_2010, @szpiech_long_2013.\n",
    "\n",
    "Here you'll work with human genotyping-microarray files from the\n",
    "[*Personal Genome Project*](http://www.personalgenomes.org/)\n",
    "(PGP). PGP is an effort by some notable scientists and thinkers to\n",
    "collect and publicize genetic and other health-related data. You can\n",
    "read more about the impetus for the project on the site.\n",
    "\n",
    "The genotype files have a simple structure, with one tab-delimited SNP\n",
    "genotype per row. Each row consists of an `rsid`, `chromosome`,\n",
    "`position`, and `genotype`. The rsid is a unique identifier for the\n",
    "SNP marker. The other three fields are self-explanatory.\n",
    "\n",
    "Write a function called `find_roh` that identifies the\n",
    "longest autosomal ROHs observed in a given human's genotype data. This\n",
    "function will take two command-line arguments. The first argument is the name of the\n",
    "genome data file. The second argument is\n",
    "the number of ROHs to report. What the function should return is\n",
    "illustrated below.\n",
    "\n",
    "You should filter out ungenotyped SNPs; those will show genotype\n",
    "`--`. You should probably proceed one chromosome at a time, and you\n",
    "should not assume that the markers are sorted. That is, sort the\n",
    "markers by position. You'll need to look at chromosomes 1 through 22\n",
    "inclusive.\n",
    "\n",
    "Once that's done, you'll want to look for the actual runs\n",
    "themselves. Here's a stretch of consecutive markers from chromosome 1:\n",
    "\n",
    "```\n",
    "rs4970383       1       838555  AC\n",
    "rs4475691       1       846808  CC\n",
    "rs7537756       1       854250  AA\n",
    "rs13302982      1       861808  GG\n",
    "rs55678698      1       864490  CT {Heterozygote that stopped the last ROH.}\n",
    "i6019299        1       871267  CC (Start of ROH.)\n",
    "rs1110052       1       873558  TT\n",
    "rs147226614     1       878697  GG\n",
    "i6052728        1       878697  GG\n",
    "i6019302        1       881843  GG\n",
    "rs2272756       1       882033  GG ... 11 consecutive homozygotes\n",
    "rs67274836      1       884767  GG\n",
    "i6019303        1       888554  CC\n",
    "rs13302945      1       889159  CC\n",
    "i6019304        1       889182  GG\n",
    "i6019305        1       891343  GG (End of ROH.)\n",
    "rs13303106      1       891945  AG {Heterozygote that stopped the ROH.}\n",
    "i6019306        1       894379  GG\n",
    "rs13303010      1       894573  AG\n",
    "i6019308        1       897792  CC\n",
    "```\n",
    "\n",
    "So, if this were among the longest ROHs in the whole genome, we would\n",
    "create a record like this:\n",
    "\n",
    "```\n",
    "{'chromosome':'1', 'start':871267, 'end':891343, 'length':11]\n",
    "```\n",
    "\n",
    "Meaning, on chromosome 1, starting at position 871267 and ending at\n",
    "position 891343, there was a run of 11 consecutive homozygotes.\n",
    "\n",
    "Your function should store the chromosome, start and stop position, and\n",
    "number of markers in each ROH as above in a dictionary. The function should\n",
    "return a list that contains the longest ROHs you found, in terms of\n",
    "the number of markers within the ROH, up to the number\n",
    "requested.\n",
    "\n",
    "#### Example\n",
    "\n",
    "This example uses data from the anonymous PGP human individual\n",
    "[`huAB47EB`](https://my.pgp-hms.org/profile/huAB47EB). I downloaded\n",
    "and unzipped her 23andMe genotype data (*her?* -- yep, her Y genotypes\n",
    "are all no-calls...) from PGP,\n",
    "[here](https://my.pgp-hms.org/user_file/download/1061), which yielded\n",
    "a simple albeit long text file,\n",
    "`genome_Lindgren_Rider_20140216145128.txt`.\n",
    "\n",
    "```\n",
    "> find_roh('genome_Lindgren_Rider_20140216145128.txt', 5)\n",
    "[{'chrom': '16', 'end': 2223985, 'length': 585, 'start': 2085997},\n",
    " {'chrom': '17', 'end': 41636310, 'length': 481, 'start': 41085683},\n",
    " {'chrom': '7', 'end': 117305628, 'length': 470, 'start': 117235056},\n",
    " {'chrom': '17', 'end': 48274031, 'length': 402, 'start': 48111137},\n",
    " {'chrom': '7', 'end': 99827148, 'length': 330, 'start': 98788589}]\n",
    "```\n",
    "\n",
    "That is, I asked for the **five** longest ROHs in huAB47EB's\n",
    "genome. The longest one was in a region on chromosome 16, from about\n",
    "2.1Mb in to 2.2Mb, spanning 585 genotyped SNPs. Meaning, huAB47EB\n",
    "appears to have a stretch in her genome of more than 100kb without a\n",
    "heterozygote!\n",
    "\n",
    "Feel free to download some other individuals from PGP, perhaps getting\n",
    "some folks of different ethnicities, to see if that length is par for\n",
    "the course. If you happen to have been genotyped yourself, you can\n",
    "download your own data and study yourself!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Solutions to Problem 1 here.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem 2\n",
    "\n",
    "During most of the time period from 2000 to 2010 many, maybe even\n",
    "most, of the world's human genetics researchers were engaged in a kind\n",
    "of genetics gold rush -- a GWAS-rush. A GWAS, or genomewide\n",
    "association study, searches for associations between some phenotype of\n",
    "interest -- breast cancer, schizophrenia, body mass index -- and\n",
    "genotyped markers, typically many hundreds of thousands of them. You\n",
    "need some number of *cases*, people/organisms having the condition,\n",
    "and some number of *controls*, those that don't. The signature that\n",
    "you're looking for is an *association* -- a correlation between having\n",
    "the condition and having some allele.\n",
    "\n",
    "In this homework, you'll take a cloud-level view at the fruits of all\n",
    "this GWAS research. The National Human Genome Research Institute, or\n",
    "NHGRI, is the part of the NIH that funds human genomic research. NHGRI worked\n",
    "together with their European counterpart, EMBL, to\n",
    "maintain a central repository of all the genetic associations\n",
    "discovered through GWAS. (There's a similar resource, FYI, for\n",
    "Mendelian disorders: the Online Mendelian Inheritance in Man, or\n",
    "OMIM.) You'll want to follow the links from EMBL's GWAS\n",
    "[catalog page](https://www.ebi.ac.uk/gwas/), to download the full\n",
    "catalog as a tab-delimited file. Follow the 'Download' link,\n",
    "and choose the file that goes with \"The most recent edition of all the associations in the GWAS Catalog\".\n",
    "It's pretty big, at 24,862 lines by my count, and\n",
    "summarizes many years and many millions of dollars of work.\n",
    "\n",
    "Each of these 25K lines, apart from the header, corresponds to a\n",
    "phenotype-genotype correlation that was discovered in a peer-reviewed\n",
    "GWAS paper. For each question below, you need to (1)\n",
    "make a matplotlib plot in your notebook, (2) write a single sentence\n",
    "summarizing any pattern that you see, even if seems obvious, and (3)\n",
    "write one further sentence commenting on the meaning of this plot, the\n",
    "*why* behind the pattern you're seeing.\n",
    "\n",
    "1. How has the number of associations reported changed year-to-year?\n",
    "   Here you need to look at how many associations there are for 2008,\n",
    "   for 2009, and so on for each of the years present in the file, and\n",
    "   make a reasonable plot with them.\n",
    "\n",
    "1. What is the distribution of risk allele frequencies for these\n",
    "   associations? Here you'll probably want to use a histogram. Allele\n",
    "   frequencies, being frequencies, should be numbers between 0\n",
    "   and 1. Not all the fields in here are numbers. Sometimes it's \"NR\",\n",
    "   \"not reported\" (*Really? Then how'd they get published?*) -- ignore\n",
    "   those. Sometimes there's both a number and a word -- in that case\n",
    "   ignore the word and just use the number.\n",
    "\n",
    "1. What is the distribution of \"Initial Sample Size\"s across these\n",
    "   studies? Here you'll probably want to plot a histogram. You'll note\n",
    "   that many of the fields in sample size are repeated, just because a\n",
    "   single study can report multiple markers. Therefore, please only\n",
    "   use one instance of any repeated entries from this column. You'll\n",
    "   also notice that many of these entries are a mix of text and\n",
    "   numbers, like \"927 North Indian ancestry cases, 1,148 North Indian\n",
    "   ancestry controls\". You want to extract the numbers from such lines\n",
    "   , and add them up. Do mind the commas. In this example, you'll pop\n",
    "   out that 927 and that 1148, and have python add them together to\n",
    "   get a total sample size of 2075.\n",
    "\n",
    "1. With a bar chart, show the top ten phenotypes (\"Disease/Trait\"), in\n",
    "   terms of the number of associations reported for them.\n",
    "\n",
    "1. With a bar chart, show the top ten cancer phenotypes, in terms of\n",
    "   the number of associations reported for them. So, subset to the\n",
    "   phenotypes that contain the string \"cancer\".\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Solutions to Problem 2.\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
