{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bio 302 Homework 10\n",
    "\n",
    "**Due before class Thursday, April 21st**\n",
    "\n",
    "This homework consists of the two problems in this notebook. You may find the reading in Chapter 10 useful. background.\n",
    "\n",
    "**Submission**\n",
    "\n",
    "Submit your solutions as a link to a Gist via a Slack DM."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 1\n",
    "\n",
    "There is a group of diseases you want nothing do with, called\n",
    "Inflammatory Bowel Disease, or IBD. Symptoms include diarrhea,\n",
    "gastrointestinal bleeding, vomiting, and abdominal cramps, among\n",
    "others. IBD is general term that comprises a number of separate\n",
    "conditions. The two major forms, which are what we'll look at here,\n",
    "are *Crohn's Disease* and *ulcerative colitis* (UC).\n",
    "\n",
    "The data set we analyze here contains measurements of the expression\n",
    "levels of 22,283 genes in peripheral blood mononuclear cells\n",
    "(PBMCs). Data for 127 subjects are included, 26 of whom have\n",
    "ulcerative colitis, and 59 of whom have Crohn's disease (the remaining\n",
    "subjects are healthy and will not be considered here). The goal is to\n",
    "identify genes that have different mean expression levels in the two\n",
    "disease groups.\n",
    "\n",
    "The raw data are available here as accession number\n",
    "[GDS1615](http://www.ncbi.nlm.nih.gov/sites/GDSbrowser?acc=GDS1615)\n",
    "from the NCBI's GEO (Gene Expression Omnibus) site. Download the\n",
    "\"DataSet SOFT file\", as opposed to the \"DataSet full SOFT file\". If\n",
    "you've got the right file, its name should be `GD1615.soft.gz`, and it\n",
    "should be 10353739 bytes. That file is compressed with `gzip`. You can\n",
    "uncompress the file with `gunzip` on macs, and with 7-zip or such on\n",
    "windows.\n",
    "\n",
    "You can compute the mean expression level of a gene in the Crohn's\n",
    "patients and UC patients, but if they differed, you wouldn't know\n",
    "whether the difference was real.\n",
    "\n",
    "One common way to approach this problem, figuring out whether two\n",
    "means are different, is via a *t-statistic*. The idea is, you look at the\n",
    "variation in expression levels within Crohn's patients and UC\n",
    "patients, respectively, to give you an idea of whether the difference\n",
    "between your means seems big or not. E.g., if the difference between\n",
    "expression levels is 50, and the standard deviations in Crohn's and UC\n",
    "patients are 4 and 7, respectively, then that looks like a real\n",
    "difference because the two distributions don't overlap much.\n",
    "\n",
    "Okay. Here's the definition of a t-statistic for the difference between\n",
    "two groups $A$ and $B$:\n",
    "\n",
    "$$\n",
    "t = \\frac{\\mu_{A} - \\mu_{B}}{\\sqrt{\\frac{\\sigma^2_{A}}{n_{A}} + \\frac{\\sigma^2_{B}}{n_{B}}}}\n",
    "$$\n",
    "\n",
    "That $\\mu$ is the mean of whatever we're measuring for the indicated\n",
    "group. The $\\sigma^2$ is the variance, a standard measure of how\n",
    "spread-out the data is. The $n$ is the sample size. Here are\n",
    "definitions for the mean and the variance (this isn't a stats class):\n",
    "\n",
    "```python\n",
    "def mean(sample):\n",
    "\t\"Compute the average of a list of numbers.\"\n",
    "\treturn sum(sample) / float(len(sample))\n",
    "\n",
    "def variance(sample):\n",
    "\t\"Compute the (sample) variance of a list of numbers.\"\n",
    "\tmn = mean(sample)\n",
    "\treturn sum((n - mn) ** 2 for n in sample) / float(len(sample) - 1)\n",
    "```\n",
    "\n",
    "An example: suppose that we had 12 males and 14 females. The mean\n",
    "height of the males is 6.0 feet and the mean height of the females is\n",
    "5.7 feet. The variance for the males is 0.0225 and for the females is\n",
    "0.01 (corresponding to standard deviations of 0.15 feet and 0.1 feet,\n",
    "respectively). Are the heights of these two groups different, or not?\n",
    "\n",
    "Plug. Then chug:\n",
    "\n",
    "$$\n",
    "t = \\frac{\\mu_{♂} - \\mu_{♀}}{\\sqrt{\\frac{\\sigma^2_{♂}}{n_{♂}} + \\frac{\\sigma^2_{♀}}{n_{♀}}}} = \\frac{6.0 - 5.7}{\\sqrt{\\frac{0.0225}{12} + \\frac{0.01}{14}}} \\doteq 5.9\n",
    "$$\n",
    "\n",
    "A t-statistic of zero means, naturally, that there's no difference\n",
    "between the group means. The farther the score gets away from zero,\n",
    "the more you tend to believe that the means are really different. The\n",
    "virtue of that fancy denominator is that the units of the t-statistic\n",
    "are independent of what you measured. A t-statistic of 1 is sort of\n",
    "unlikely if the means are the same, a t-statistic of 2 is getting\n",
    "really unlikely that the means are the same, and a t-statistic of 3\n",
    "means there is almost no chance that the means are the same. A\n",
    "t-statistic of 5.9 is getting pretty darned close to certainty. So for\n",
    "these data that's strong evidence that men and women aren't the same\n",
    "height.\n",
    "\n",
    "*Example* Here's the top-left corner of the data table from expression\n",
    "dataset. `ID_REF` is some unique id, `IDENTIFIER` is the gene name,\n",
    "and `GSM76115` and `GSM76116` are anonymous ids for two of the human\n",
    "beings that participated in the study. You can see from the metadata\n",
    "at the top of the file that those two people were in the *normal*\n",
    "group, and not in the UC or Crohn's groups. *Note: these columns are\n",
    "tab-delimited in the actual file. I spaced them out as below for\n",
    "legibility.*\n",
    "\n",
    "~~~\n",
    "   ID_REF     IDENTIFIER  GSM76115  GSM76116\n",
    "   1007_s_at        DDR1   80.7287   62.9846\n",
    "   1053_at          RFC2   25.9926   22.6722\n",
    "~~~\n",
    "\n",
    "The expression level of the gene *DDR1*, in the individual `GSM76115`,\n",
    "was 80.7287. And so on.\n",
    "\n",
    "\n",
    "1. Plot the distributions of the log-transformed expression levels\n",
    "   across all of the genes, for UC and Crohn's patients as two\n",
    "   separate histograms. \"Log-transformed\" means to work with the\n",
    "   logarithm of the values, instead of the values themselves. Use the\n",
    "   `log10` function provided with python's `math` module. You can\n",
    "   obtain the identifiers for the three subsets readily from the\n",
    "   `subset_sample_id` lines in the file header.\n",
    "\n",
    "1. Compute the t-statistics for the log-transformed expression levels\n",
    "   for each gene in the dataset, between UC patients and Crohn's\n",
    "   patients. Show the genes, and their t-statistics, for the 20 genes\n",
    "   with the t-statistics farthest from zero, so those with the most\n",
    "   evidence that their mean expression levels are not the same. You'll\n",
    "   probably find the built-in `abs` function helpful.\n",
    "\n",
    "1. Why would we be interested in these genes, the ones with the high\n",
    "   t-statistics? Explain.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Solutions to Problem 1 here.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2\n",
    "\n",
    "These questions refer to Diehn *et al* (2005), which is available\n",
    "in html and pdf form [here](http://genomebiology.com/2005/6/9/R74/).\n",
    "\n",
    "1. Read enough of the paper so that you understand what's going on in\n",
    "   Figure 1. Then, please rewrite the figure legend in your own words.\n",
    "\n",
    "1. It's fascinating to read their summaries of each tissue type in the\n",
    "   eye. To pick one, please read the *Lens signature* section, and\n",
    "   summarize its findings in your own words.\n",
    "\n",
    "1. In the section *Identifying candidate disease genes*, the authors\n",
    "   combine expression data with disease information in order to, well,\n",
    "   identify potential eye-related disease genes. Summarize the\n",
    "   reasoning and evidence behind this section.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Solutions to Problem 2 here.\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
